lib/Server.js

```diff
@@ -1931,7 +1931,29 @@ class Server {
       const _path = req.path;
 
       try {
-        const filename = this.middleware.getFilenameFromUrl(`${_path}.js`);
+        const compilerOptions = this.getCompilerOptions();
+        const isOutputModuleEnabled =
+          compilerOptions.experiments &&
+          compilerOptions.experiments.outputModule;
```

Comment:
Developer can set `js` extension for modules too, ideally we should look at info for assets, i.e. compilation.getAssetInfo and check on `module: true | false | undefined`